src/serializer/modules.js

```diff
@@ -70,64 +76,88 @@ class ModuleTracer extends Tracer {
+        this.moduleIdsRequiredInEvaluateForEffects.add(moduleIdValue);
+      } else {
+        let result;
+        try {
+          this.requireStack.push(moduleIdValue);
+          let requireSequenceStart = this.requireSequence.length;
+          this.requireSequence.push(moduleIdValue);
+          let anyRequiresAccelerated, effects;
+          do {
+            effects = realm.evaluateForEffects(() => {
+              try {
+                return performCall();
+              } catch (e) {
+                if (e instanceof Completion) return e;
+                throw e;
+              }
+            }, this);
+
+            // We gathered all effects, but didn't apply them yet.
+            // Let's check if there was any call to `require` in a
+            // evaluate-for-effects context. If so, try to initialize
+            // that module right now.
+            anyRequiresAccelerated = false;
+            for (let moduleIdRequiredInEvaluateForEffects of this.moduleIdsRequiredInEvaluateForEffects) {
+              let acceleratedEffects = this.modules.tryInitializeModule(
+                moduleIdRequiredInEvaluateForEffects,
+                `accelerated initialization of module ${moduleIdRequiredInEvaluateForEffects} as it's required in an evaluate-for-effects context by module ${moduleIdValue}`);
+              if (acceleratedEffects === undefined || !(acceleratedEffects[0] instanceof IntrospectionThrowCompletion)) {
+                anyRequiresAccelerated = true;
+              }
+            }
+            this.moduleIdsRequiredInEvaluateForEffects.clear();
+            // Keep restarting for as long as we find additional modules to accelerate.
+          } while (anyRequiresAccelerated);
+
+          [result] = effects;
+          invariant(result instanceof Value || result instanceof Completion);
+          if (result instanceof IntrospectionThrowCompletion) {
+            let [message, stack] = this.modules.getMessageAndStack(effects);
+            console.log(`delaying require(${moduleIdValue}): ${message} ${stack}`);
+            // So we are about to emit a delayed require(...) call.
+            // However, before we do that, let's try to require all modules that we
+            // know this delayed require call will require.
+            // This way, we ensure that those modules will be fully initialized
+            // before the require call executes.
+            // TODO: More needs to be done to make the delayUnsupportedRequires
+            // feature completely safe. Open issues are:
+            // 1) Side-effects on the heap of delayed factory functions are not discovered or rejected.
+            // 2) While we do process an appropriate list of transitively required modules here,
+            //    more modules would have been required if the Introspection exception had not been thrown.
+            //    To be correct, those modules would have to be prepacked here as well.
```

Comment:
The evaluator will not reliably discover all require calls because of a loss of precision. We could alternatively use a specialized visitor that uses pattern matching to find the calls in a flow insensitive way. The downside of that is that we would miss require calls that happen via aliases. 

Perhaps the best way is to always do the simple visitor, and then the evaluator and to produce an error message whenever the evaluator finds a require call that was missed by the simple visitor. I.e. lets require and try to enforce, the requirement that require calls are discoverable by the simple visitor.